File: 2024/js/day01/pt02.js
const fs = require('fs')

function calculateSimilarityScore(leftList, rightList) {
    // Create a frequency map for the right list
    const frequencyMap = {}
    rightList.forEach(num => {
        frequencyMap[num] = (frequencyMap[num] || 0) + 1
    })

    // Calculate the similarity score
    let similarityScore = 0
    leftList.forEach(num => {
        if (frequencyMap[num]) {
            similarityScore += num * frequencyMap[num]
        }
    })

    return similarityScore;
}

function readInputFile(filePath) {
    const data = fs.readFileSync(filePath, 'utf8')
    const leftList = []
    const rightList = []
    
    // Parse the input file line by line
    data.trim().split('\n').forEach(line => {
        const [left, right] = line.split(/\s+/).map(Number)
        leftList.push(left)
        rightList.push(right)
    });

    return { leftList, rightList }
}

function main() {
    // const filePath = '../../inputs/day01/testInput.txt'
    const filePath = '../../inputs/day01/input.txt'
    const { leftList, rightList } = readInputFile(filePath)
    const similarityScore = calculateSimilarityScore(leftList, rightList)
    console.log('Similarity Score:', similarityScore)
}

main()

File: 2024/python/day01/pt02.py
def calculate_similarity_score(left_list, right_list):
    # Create a frequency dictionary for the right list
    frequency_map = {}
    for num in right_list:
        frequency_map[num] = frequency_map.get(num, 0) + 1

    # Calculate the similarity score
    similarity_score = 0
    for num in left_list:
        if num in frequency_map:
            similarity_score += num * frequency_map[num]

    return similarity_score


def read_input_file(file_path):
    left_list = []
    right_list = []

    # Read and parse the file
    with open(file_path, 'r') as file:
        for line in file:
            left, right = map(int, line.split())
            left_list.append(left)
            right_list.append(right)

    return left_list, right_list


def main():
    # file_path = '../../inputs/day01/testInput.txt'
    file_path = '../../inputs/day01/input.txt'
    left_list, right_list = read_input_file(file_path)
    similarity_score = calculate_similarity_score(left_list, right_list)
    print('Similarity Score:', similarity_score)


if __name__ == '__main__':
    main()
